### PR TITLE
ci: add a rule-probe script, and record the shared path-insensitivity

### PR DESCRIPTION
**Pull request by ariwk**
Two small additions that came out of #101, now that it is merged.

## `tests/probe_rule.sh`

`test_rules.sh` answers whether the rules still match their fixtures. It does not answer what a rule does to one particular file, which is the question every rule change raises before a fixture exists for it — the shape a review comment describes, a hardening spelling nobody has pinned. Seven review rounds on #101 were spent hand-rolling that probe, once per shape, in a scratch script that does not survive the session.

What makes it worth a checked-in file rather than a remembered semgrep invocation is the failure mode it closes. An invalid rule file makes semgrep load no rule at all from a `--config` directory and exit 7 with both streams empty under `--quiet`, so a hand-rolled probe that ignores the exit status prints `0 findings` for a rule that never ran — indistinguishable from a rule that ran and matched nothing, and the trap this repository's own `CLAUDE.md` already documents. The script checks the status, surfaces `.errors[]`, and refuses to print a count when semgrep exited non-zero.

Verified in both directions against a deliberately invalid rule file:

```
=== …/velocity-ssti.vuln.java ===
SEMGREP EXIT 7 — no finding count below is meaningful
  ERROR error: Invalid YAML file …
  ERROR error: invalid configuration file found (1 configs were invalid)
tool exit=1
```

and on a valid rule, the five asserted cases of the Velocity fixture with their line numbers.

Worth recording, since it cost a round here: the first attempt at that negative test produced a *valid* file — an undoubled quote inside a **double**-quoted YAML scalar is legal, and the `CLAUDE.md` trap is about single-quoted ones. It loaded fine and reported six findings, which reads exactly like a passing test of the wrong thing.

## The velocity gap entry

#101 recorded that hardening behind an `if`, or written in a lambda that is never invoked, clears `polarion-xxe-unsafe-parser` although nothing is applied on the path that reaches the parser. `polarion-velocity-ssti` branch A has the same property and the entry did not say so. Measured on semgrep 1.172.0 — both shapes report 0 — so the entry now points at the one below it rather than leaving the asymmetry for a triager to discover.

No rule behaviour changes here: the suite passes 8/8 unchanged, velocity 5/5 and XXE 9/9 with both fixed fixtures at 0.

Refs #101

### CODE REVIEW

File: polarion-semgrep-rules/tests/probe_rule.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    # reassuring 0, which is the whole trap this script exists to close.
+    if [[ ${rc} -ne 0 ]]; then
+        printf 'SEMGREP EXIT %s — no finding count below is meaningful\n' "${rc}"
+        printf '%s' "${json}" | jq -r '.errors[]? | "  ERROR \(.level): \((.message // "") | split("\n")[0])"' 2>/dev/null || true
+        status=1
+        continue
+    fi
+
+    printf '%s' "${json}" | jq -r '.errors[]? | "  ERROR \(.level): \((.message // "") | split("\n")[0])"'
+    printf 'count=%s\n' "$(printf '%s' "${json}" | jq '.results | length')"
```
[the remark this answers is not in the document]
Comment (ariwk) [reply polarion-semgrep-rules/tests/probe_rule.sh]: Confirmed by measurement and fixed in 895622b. Reproduced the exact case against this repository's own fixtures on semgrep 1.172.0: `probe_rule.sh rules tests/fixtures` printed count=0 with .paths.scanned empty and exit 0, because the built-in .semgrepignore skips tests/ — the fixture directory an author would reach for first. The probe now prints `count=N (scanned M file(s))` and, when M is zero, an explicit note that the count says nothing about the code. The header comment now names both doors into the silent zero rather than only the exit-7 one.

File: polarion-semgrep-rules/tests/probe_rule.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    if [[ ! -e "${target}" ]]; then
+        echo "FAIL: ${target} does not exist" >&2
+        status=1
+        continue
+    fi
+
+    printf '=== %s ===\n' "${target}"
+
+    set +e
+    json=$("${SEMGREP_CMD[@]}" --metrics off --json --quiet --config "${rule}" "${target}")
```
[the remark this answers is not in the document]
Comment (ariwk) [reply polarion-semgrep-rules/tests/probe_rule.sh]: Confirmed and fixed in 895622b — the flag is now passed, and the header says why. One correction to the reasoning, because it matters to anyone who reads this thread later: the rewrite is driven by the --config PATH, not by the config being a directory. Measured on 1.172.0, `--config rules/velocity-ssti.yaml` — a single rule file — also reports rules.polarion-velocity-ssti. So both of this script's modes needed the flag, not just the whole-pack one. test_rules.sh is unaffected for a different reason than the one given: it counts .results | length and never prints a check_id.

File: polarion-semgrep-rules/tests/probe_rule.sh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    # A count is only an answer about code semgrep actually read, so print the
+    # scanned-file count beside it and say so outright when it is zero.
+    scanned=$(printf '%s' "${json}" | jq '.paths.scanned | length')
+    printf 'count=%s (scanned %s file(s))\n' \
+        "$(printf '%s' "${json}" | jq '.results | length')" "${scanned}"
+    if [[ "${scanned}" -eq 0 ]]; then
+        printf '  NOTE: semgrep scanned nothing here — the extension does not map to the\n'
+        printf '        rule language, or the path is skipped by .semgrepignore. The count\n'
+        printf '        above says nothing about this code.\n'
+    fi
```
[the remark this answers is not in the document]
Comment (ariwk) [reply polarion-semgrep-rules/tests/probe_rule.sh:93]: Agreed and fixed in c1523d4 — `status=1` on that branch, so the two doors now behave the same way rather than one printing a warning and the other failing. Measured on 1.172.0 across four paths: normal target exit 0, `probe_rule.sh rules tests/fixtures` exit 1 with the note, a mixed target list exit 1 (status is shared across targets, as it already is for a missing target), invalid rule file still exit 1 with no count. Your scoping is right and I am not widening it here: a directory where some files scan and others are skipped keeps scanned > 0, prints no note and exits 0. Closing the total-zero case is what the exit-7 branch's symmetry buys; per-file skip accounting would be a different feature.

File: CLAUDE.md [hunk 4]
```diff
@@ -23,6 +23,7 @@ This repository contains **GitHub Actions workflows (reusable and caller/CI)** a
...
 - Verify a rule-pack change with `SEMGREP="uvx semgrep==$(cat polarion-semgrep-rules/SEMGREP_VERSION)" bash polarion-semgrep-rules/tests/test_rules.sh` — without the `SEMGREP` override the suite runs whatever semgrep is on PATH, which is not the version the rule baseline was measured on
 - Re-measure the corpus with `POLARION_TARGETS_DIR=<dir> bash polarion-semgrep-rules/tests/measure_corpus.sh` rather than by hand — it pins the semgrep version and the excludes that decide whether two measurements are comparable, prints the per-rule breakdown so a total that held while its composition changed is visible, and fails on a missing target instead of skipping it
+- Probe a rule against arbitrary code with `bash polarion-semgrep-rules/tests/probe_rule.sh <rule|rules-dir> <target>…` rather than calling semgrep by hand — it pins the version, prints the per-file count with line numbers, and refuses to print a count at all when semgrep exited non-zero, which is the exit-7 case below reaching you as a failure instead of as a reassuring 0
```
[the remark this answers is not in the document]
Comment (ariwk) [reply CLAUDE.md]: Agreed and fixed in c1523d4. The argument that carried it is the one you gave: CLAUDE.md loads in every session and already records two bullets down that the built-in .semgrepignore skips tests/, so the two lines combined into the trap rather than either being wrong alone. The bullet now covers both doors, names the tests/fixtures/ probe as the concrete case, and records that the scanned-zero path exits 1. I shortened your suggested wording rather than taking it verbatim — same two facts, kept to one entry without restating the .semgrepignore mechanism the fixtures bullet already carries.
